dll/win32/kernel32/client/handle.c

```diff
@@ -187,6 +221,134 @@ DuplicateHandle(IN HANDLE hSourceProcessHandle,
         return FALSE;
     }
 
+    // winsock socket handles code currently only dupes from self or inherited sockets from parent are implemented
+    // TODO: add socket duplication from arbitrary source process (works in Windows 2003)
+    WCHAR nameFull[80];
+    ULONG returnedLength;
+    BOOLEAN isSocket = FALSE;
+    BOOLEAN canDupe = FALSE;
+    _SEH2_TRY
+    {
+        // we rely on a side effect of a failed call.. we get the name even though call failed
+        NtQueryObject(hSourceHandle, ObjectNameInformation, nameFull, sizeof(nameFull), &returnedLength);
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        int excd = _SEH2_GetExceptionCode();
+        DPRINT1("DuplicateHandle: NtQueryObject generated an exception (%lx) for handle(%lx)\n", excd, hSourceHandle);			
+    }
+    _SEH2_END;
+    if (wcscmp((LPWSTR)(nameFull + (2 * sizeof(WCHAR))), L"\\Device\\Afd") == 0)
+    { // check if this belongs to AFD, if so it's a winsock socket
+        DPRINT("DuplicateHandle: NtQueryObject winsock socket device(AFD) detected for handle(%lx)\n", hSourceHandle);				
+        // DynLink the winsock routines we need for this work.
+        DPRINT("DuplicateHandle: calling GetModuleHandleA ws2_32\n");
+        winsock_module = GetModuleHandleA("ws2_32");
+        if (winsock_module)
+        {
+            DynWSASocket = (wsskcall)GetProcAddress(winsock_module, "WSASocketW");
+            getsockopt = (wsgsocall)GetProcAddress(winsock_module, "getsockopt");
+            if (getsockopt) 
+            {
+                // getsockopt is our gatekeeper. if it fails winsock access is not in a reliable state
+                int result = getsockopt((SOCKET)hSourceHandle, SOL_SOCKET, SO_TYPE, (char*)&optVal, &optLen);
+                if (result == 0) 
+                {
+                    if(optVal == SOCK_STREAM)
+                        DPRINT("(PID %lx) DuplicateHandle: socket is a TCP socket\n");
+                    else if(optVal == SOCK_DGRAM)
+                        DPRINT("DuplicateHandle: socket is a UDP socket\n");
```

Comment:
Is `if(optVal == SOCK_DGRAM)` superfluous?
Or would we want an additional `else` case?

Reply:
Removed whole section, it was mostly traces for testing.